blog/2016-04-20-futhark-and-pygame.rst

```diff
@@ -0,0 +1,215 @@
+          if event.type == pygame.QUIT:
+              sys.exit()
+
+The ``steps`` argument is the number of simulation steps to perform
+per frame, and defaults to 3.  You can set this to any positive 32-bit
+int.  To increase the work done per frame, the default is 3 and not 1.
+
+We have also added a simple PyGame event check, so that you can close
+the simulation window as expected.
+
+
+The Three Other Ones
+--------------------
+
+In the fluid simulator you can add both particles and forces.  See for yourself:
+
+.. raw:: html
+
+   <video src="/static/fluid-2016.04.20.webm"
+          poster="/static/fluid-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/fluid-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/fluid-2016.04.20.webm">fluid-2016.04.20.webm</a></p>
+
+My GPU is not the best one around, so I am running this in a fairly
+small window.
+
+The Mandelbrot Explorer is also pretty nifty:
+
+.. raw:: html
+
+   <video src="/static/mandelbrot-2016.04.20.webm"
+          poster="/static/mandelbrot-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/mandelbrot-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/mandelbrot-2016.04.20.webm">mandelbrot-2016.04.20.webm</a></p>
+
+In the end of the video, I switch to NumPy's internal Mandelbrot generator.
+
+Finally, there is the HotSpot 2D Heat Equation GUI.  You can see its
+performance `here </performance.html#hotspot-futhark-rodinia>`_.  This
+visualisation is pretty silly.
+
+.. raw:: html
+
+   <video src="/static/hotspot-2016.04.20.webm"
+          poster="/static/hotspot-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/hotspot-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/hotspot-2016.04.20.webm">hotspot-2016.04.20.webm</a></p>
+
+
+Use Cases
+---------
+
+Futhark is an optimising compiler which takes an *entire program* as
+input.  As such, its optimisations are not directed at separate
+functions, but rather the program as a whole.  This is important to
+keep in mind when developing Futhark-PyGame programs, since it means
+that we would like to have as few calls as possible to Futhark from
+Python, and keep as much code as possible inside Futhark.
```

Comment:
While everything here is true, the conclusion does not follow from the premise.  The reason one should minimise calls is because Futhark copies everything back and forth between CPU and GPU for every call.